### PR TITLE
Input delivery information

### PR DESCRIPTION
**Pull request by RidicZhi**
finish the functionality of DetailsForm and each input validation.
add 2 test files for component DetailsForm and its child - Input.

### CODE REVIEW

**Review by raytf — APPROVED** [review]

File: src/components/App.js [hunk 1]
```diff
@@ -1,20 +1,50 @@
...
+
+  onDetailsFormDataChange(name, value) {
+    const { detailsFormData } = this.state;
+
+    const newDetailsFormData = {
+      ...detailsFormData,
+      [name]: value,
+    };
+
+    this.setState({
```
Comment (KieraDOG) [inline src/components/App.js:32]:
```
this.setState((state) => {
const { detailsFormData } = this.state;

      const newDetailsFormData = {
       ...detailsFormData,
       [name]: value,
     };

return { detailsFormData: newDetailsFormData }
})
```

File: src/components/DetailsForm.js [hunk 2]
```diff
@@ -1,27 +1,133 @@
...
+          onChange={value => onDataChange('postCode', value)}
+          validationMessage={getValidationMessageForPostCode(data.postCode)}
+        />
+      </DetailsFormDiv>
+    );
+  }
+}
+
+function validatePostCode(postCode) {
+  var re = /^\d{4}$/;
```
Comment (KieraDOG) [inline src/components/DetailsForm.js:73]: Why do we need `var` here?

File: src/components/DetailsForm.js [hunk 3]
```diff
@@ -1,27 +1,133 @@
...
+
+  if (!validatePostCode(postCode)) {
+    return 'The Post Code format is invalid';
+  }
+
+  return '';
+}
+
+function validateEmail(email) {
+  var re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
```
Comment (KieraDOG) [inline src/components/DetailsForm.js:90]: `regex` is better than `re`

File: src/components/Input.js [hunk 4]
```diff
@@ -25,11 +28,57 @@ const Input = styled.input`
...
+
+    const {
+      dirty,
+    } = this.state;
+
+    const showError = !validate && (dirty || formDirty);
+
+    return (
+      <Container>
+        <Label className={showError && 'error'}>{label}</Label>
```
Comment (KieraDOG) [inline src/components/Input.js:74]: <Label showError={showError}>{label}</Label>

const Label = styled.div`
${({ showError }) => showError && `color: red; ...`}
`;

File: src/components/Input.js [hunk 5]
```diff
@@ -25,11 +28,57 @@ const Input = styled.input`
...
+        <InputFrame
+          value={value}
+          onChange={this.onChange}
+          type="text"
+        />
+        <ValidationErrorDiv>{showError && validationMessage}</ValidationErrorDiv>
```
Comment (KieraDOG) [inline src/components/Input.js:80]: {showError && (<ValidationErrorDiv>{validationMessage}</ValidationErrorDiv>)}

**Pull request merged**